Support methods from external classes in @MethodSource #1088

karollewandowski opened the pull request:
I cleaned by the way unused `throws` clauses in `ReflectionUtilsTests`.
FQMN parsing result is just a String array, I think it's simple enough and creating something like `MethodInfo` data class would be too much.
Please carefully review added test cases - I didn't have too much time for it and could miss some important ones.

sbrannen commented:
Please add `Issue: #1088` to the bottom of each commit message.

sbrannen commented:
> I cleaned by the way unused throws clauses in ReflectionUtilsTests.

Would you mind doing one of the following?

- submitting a separate PR for that
- moving those changes to a separate commit within the existing PR
- allowing one of the team to make those changes on `master`

karollewandowski commented:
@sbrannen, I squashed everything to single commit and reverted cleaning unnecessary `throws` clauses. I noticed there is no convention to have it everywhere and just wanted to be a good scout ;-) I'm leaving it for the team.

karollewandowski commented:
I just realized that I forgot to include new changes in docs. Should I do it or docs are updated in scope of separate issues?

sbrannen commented:
> @sbrannen, I squashed everything to single commit and reverted cleaning unnecessary throws clauses. I noticed there is no convention to have it everywhere and just wanted to be a good scout ;-) I'm leaving it for the team.

Sounds good

sbrannen commented:
> I just realized that I forgot to include new changes in docs. Should I do it or docs are updated in scope of separate issues?

Updates to the User Guide, Release Notes, and JavaDoc should be part of the PR.

karollewandowski commented:
@sbrannen, docs updated.
There is a problem with Spotlight analysis. I created `documentation/src/test/java/example/ExternalMethodSourceTestDemo.java` to demonstrate factory method from external class, but Spotlight gets violation on tag declaration:
```
...
// tag::external_MethodSource_example[]    <-- Spotlight removes it
+package example;
...
```
I want to put it before package to clearly demonstrate full FQNM with package name. I tried to put `//@formatter:off` before tag, but didn't help. Any ideas?

sbrannen commented:
> @sbrannen, docs updated.

Thanks!

We'll have a look...


> I want to put it before package to clearly demonstrate full FQNM with package name. I tried to put //@formatter:off before tag, but didn't help. Any ideas?

Yeah: it simply won't work. 😉 

Spotless requires that the copyright header be the first and only thing before the `package` declaration line in any given source file.

So the only option is to mention the package explicitly in the prose describing the example.

sbrannen commented on documentation/src/docs/asciidoc/release-notes/release-notes-5.2.0-M1.adoc:
```diff
@@ -39,7 +39,8 @@ on GitHub.
 ==== New Features and Improvements
 
-* ❓
+* `@MethodSource` now supports external classes factory methods referenced by fully
```
Please change to:

`@MethodSource` now supports `static` factory methods declared in external classes referenced by _fully qualified method name_.

sbrannen commented on documentation/src/docs/asciidoc/user-guide/writing-tests.adoc:
```diff
@@ -841,10 +841,11 @@ include::{testDir}/example/ParameterizedTestDemo.java[tags=EnumSource_regex_exam
-Such methods must return a `Stream`, `Iterable`, `Iterator`, or array of arguments. In
-addition, such methods must not accept any arguments. By default such methods must be
-`static` unless the test class is annotated with `@TestInstance(Lifecycle.PER_CLASS)`.
+`@MethodSource` allows you to refer to one or more _factory_ methods of the test or
```
"of the test or external class" --> "of the test class or external classes"

sbrannen commented on documentation/src/docs/asciidoc/user-guide/writing-tests.adoc:
```diff
@@ -841,10 +841,11 @@ include::{testDir}/example/ParameterizedTestDemo.java[tags=EnumSource_regex_exam
+external class. Such methods must return a `Stream`, `Iterable`, `Iterator`, or array
+of arguments. In addition, such methods must not accept any arguments. By default such
+methods must be `static` unless the test class is annotated with
+`@TestInstance(Lifecycle.PER_CLASS)`.
```
The last sentence is no longer technically correct.

Factory methods in external classes must always be `static`.

Thus, please reword that to take into account the change in this PR.

sbrannen commented on documentation/src/docs/asciidoc/user-guide/writing-tests.adoc:
```diff
@@ -880,6 +881,14 @@ factory method defined in the `Arguments` interface.
 ----
 
+External _factory_ method can be referenced by providing _fully qualified method name_ as
+demonstrated in the following example.
```
Please change to:

An external, `static` _factory_ method can be referenced by providing its _fully qualified method name_ as
+demonstrated in the following example.

sbrannen commented on documentation/src/test/java/example/ExternalMethodSourceTestDemo.java:
```diff
@@ -0,0 +1,34 @@
+import org.junit.jupiter.params.ParameterizedTest;
+import org.junit.jupiter.params.provider.MethodSource;
+
+class ExternalMethodSourceTests {
```
Please rename both the _file_ and the _class_ to: `ExternalMethodSourceDemo`.

Example classes in the `documentation` project must be named "*Demo".

sbrannen commented on documentation/src/test/java/example/ExternalMethodSourceTestDemo.java:
```diff
@@ -0,0 +1,34 @@
+ * http://www.eclipse.org/legal/epl-v20.html
+ */
+
+// tag::external_MethodSource_example[]
```
Please move the _tag_ declaration below the `package` declaration. See comments in PR.

sbrannen commented:
@karollewandowski, are you comfortable with the current _implementation_ in this PR?

In other words, aside from the documentation, do you feel this PR is ready for a final review?

karollewandowski commented:
> So the only option is to mention the package explicitly in the prose describing the example within the User Guide.

I inlined `package example;` just before file include in `writing-tests.adoc`. If it's too hacky, please let me know to revert it.

> I requested a few minor changes for the documentation.

All comments resolved, please take a look again.

> @karollewandowski, are you comfortable with the current implementation in this PR?

Yes, please review it.

sbrannen commented:
@karollewandowski, thanks for making changes.

I'll make sure to give it a final review and/or merge it in the next day or so.

sbrannen commented:
@karollewandowski, 

FYI: I'd hoped to get to _this_ issue today, but I just lost quite a bit of time with this super annoying 🐛: #1345

sbrannen commented:
This has been merged into `master` in f415edc742cda271f3c1f5c5e9b20247e92db87b and minorly polished in 13de8c51701b69769328b55b1857631e75bc5cdc.

Cheers!
